### PR TITLE
testutil,skills/types,skills,daemon: tweak discovery of know skill types

### PR DESCRIPTION
**Pull request by zyga**
This branch moves the discovery of known skill types from skills/ to skills/types.

### CODE REVIEW

**Comment by zyga** [comment]
Merged with master to re-test with integration test fixes

**Comment by mvo5** [comment]
There are two test failures in the integration tests that are  worth investigating

**Comment by fgimenez** [comment]
@mvo5 we have had this problem before [1] it appears eventually when trying to install a snap from the local filesystem, maybe related to the snap.yaml filled with '\0'

[1] http://pastebin.ubuntu.com/15126149/

**Comment by mvo5** [comment]
I rerun the tests and its happy again

File: testutil/checkers.go [hunk 1]
```diff
@@ -38,19 +38,32 @@ var Contains check.Checker = &containsChecker{
...
 	case reflect.Slice, reflect.Array, reflect.Map:
-		// Ensure that type of elements in container is compatible with elem
-		if elemV := reflect.ValueOf(elem); containerV.Type().Elem() != elemV.Type() {
-			*result = false
-			*error = fmt.Sprintf(
-				"container has items of type %s but expected element is a %s",
-				containerV.Type().Elem(), elemV.Type())
-			return true
+		containerElemType := containerV.Type().Elem()
+		if containerV.Type().Elem().Kind() == reflect.Interface {
```
Comment (niemeyer) [inline testutil/checkers.go]: This can be containerElemType as well, created right above.

**Comment by niemeyer** [comment]
Nice change, LGTM.

File: testutil/checkers.go (hunk 1 above)
Comment (zyga) [inline testutil/checkers.go]: Right, nice, I didn't notice that, updated.

File: daemon/daemon.go [hunk 2]
```diff
@@ -261,9 +262,10 @@ func New() *Daemon {
 		panic(err.Error())
 	}
 	skillRepo := skills.NewRepository()
-	err = skills.LoadBuiltInTypes(skillRepo)
-	if err != nil {
-		panic(err.Error())
+	for _, skillType := range types.AllTypes() {
```
Comment (mvo5) [inline daemon/daemon.go]: Nice!

File: skills/types/all_test.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	. "gopkg.in/check.v1"
+)
+
+type AllSuite struct{}
+
+var _ = Suite(&AllSuite{})
+
+func (s *AllSuite) TestAllTypes(c *C) {
+	allTypes := types.AllTypes()
+	c.Check(allTypes, Contains, &types.BoolFileType{})
```
Comment (mvo5) [inline skills/types/all_test.go]: Sorry for being "mr-smarty-pants" here: `c.Check(allTypes, DeepEquals, []skills.Type{&types.BoolFileType{}})` would  have worked as well. But now Contains knows about interfaces which is nice as well!

File: testutil/checkers_test.go [hunk 4]
```diff
@@ -111,12 +146,16 @@ func (s *CheckersS) TestContainsSlice(c *C) {
 	c.Assert([]int{1, 2, 3}, Contains, 2)
 	c.Assert([]int{1, 2, 3}, Contains, 3)
 	c.Assert([]int{1, 2, 3}, Not(Contains), 4)
+	c.Assert([]animal{&dog{}, &cat{}}, Contains, &dog{})
```
Comment (mvo5) [inline testutil/checkers_test.go]: Nice!

**Comment by mvo5** [comment]
Looks good!

**Pull request merged**